### ISSUE TITLE
YARN workflow representation

### ISSUE THREAD

**Issue by jdakka** [radical-collaboration/MSKCC#6]
@andrrizzi 

Could you take a look at the diagrammatic representation of YARN:

https://docs.google.com/drawings/d/1ds1eRqqCM9T_gEklX1a1f3I1ojx6RIk4JsgBNlDT4Bg/edit?usp=sharing

Feel free to add/modify/comment

**Comment by andrrizzi** [comment]
That looks about right!

To be more precise the "gathering" and "replica exchange" steps can be further subdivided. Currently, this is how it goes:
1. After propagation, the positions are gathered into the MPI node with rank 0. The only purpose of this gathering step is to be able to write them in the netcdf database so, depending on how we solve the writing to disk problem, it may not be necessary.
2. The energy matrix is computed in parallel. Each replica is assigned to `n_replicas / n_mpi_processes` rows of the matrix.
3. There's another gathering step to collect all the energy matrix rows into MPI node 0.
4. MPI node 0 performs Gibbs sampling using the full energy matrix and generate a new permutation of the state vector.
5. MPI node 0 sends the state vector to all MPI processes and the new states are assigned to replicas.

The order of 1) and 2) can be interchanged, if that makes it easier.

**Comment by jchodera** [comment]
I wonder if it would be more scalable (and also more scientifically interesting) to focus on SAMS instead of replica-exchange here for free energy calculations. That will greatly simplify the workflows for each ligand or mutant.

**Comment by andrrizzi** [comment]
Definitely! Depending on our time constraints though, I'm not sure I'll be able to include this into YANK before late Jan. Unless you're thinking about doing it outside YANK.

**Comment by jchodera** [comment]
Since EnsembleTk is python 2.7-only right now, it sounds like we might have to build a special 2.7-compatible code anyway if we wanted to run with EnsembleTk. Our best bet might be to try to get everything we can shoved into `openmmtools` and prepare special SAMS and Perses workflows with EnsembleTk where we can set up the systems with YANK and Perses outside of the workflow and then run the calculations using the bare minimum of code.

That would only work if most of the code we need in openmmtools is python 2.7-compatible, however. Since we switched to python 3-only, I'm not sure which parts of `openmmtools` would be OK to use in python 2.7, which is really what the feasibility of running anything with EnsembleTk depends on.

**Comment by andrrizzi** [comment]
I believe `openmmtools` is one of the few packages we have around that is still compatible with both python 2.7 and python 3 so that would definitely make things easier!

**Comment by jchodera** [comment]
Wow, you're right! We're even still testing it on python 2.7!
https://travis-ci.org/choderalab/openmmtools

OK, I think this is the best plan, then.

@jdakka : Any chance you can swing by MSKCC once more before Patrick leaves? That would at least let us come up with some minimal workflows for you to test and help us identify all the components we need to move to `openmmtools` to make this work.

**Comment by jchodera** [comment]
(I'm back in NYC on Thu and Fri this week)

**Comment by shantenujha** [comment]
@jdakka  - that would be good. 

@jchodera: apologies I won't be able to do so, as I'm in Brussels + Rejyakvik this week and early next.

**Comment by jdakka** [comment]
@andrrizzi where we left off, did you get the chance to try the RADICAL-Pilot examples? I'll update the workflow diagram based on feedback. We can work together to get YANK integrated into EnTK. I'd be fine with coming to Sloan, once you feel that we're at a good point to hack it.

**Comment by andrrizzi** [comment]
@jdakka sorry I've been crazy busy with a couple of deadlines. I should be able to go through the examples and find eventual problems that we'll have to solve by the end of next week.

**Comment by jdakka** [comment]
@andrrizzi based on the pseudo code you provided I had a question about the `compute_energy` function:`compute_energy(thermodynamic_state, sampler_state)`  Is this another MD/MC step?

**Comment by andrrizzi** [comment]
Nope! This only computes the energy of the system given the alchemical lambda parameters, which are in the `thermodynamic_state` object, and positions/box vectors in `sampler_state`.

**Comment by jdakka** [comment]
@andrrizzi I'm looking for a few more detail to understand how we can best represent Yank:

- since YANK is a one-world MPI application, and the rank is dependent on the available n_processes and m_replicas, could you roughly estimate n and m, based on a single Yank protocol instance? I assume this also heavily depends on the mutation/physical system, but could you provide an lower/upper bound?  

- also, roughly what is the number of mutations? Does each Yank protocol instance reference a different mutation?
